chapters/os_tooling.md

```diff
@@ -0,0 +1,84 @@
+
+This section describes available open source tools for offline compilation of OpenCL kernels.
+
+## Open Source Tools
+
+* [clang](https://clang.llvm.org/) is a compiler front-end for the C/C++ family of languages including OpenCL C and C++ for OpenCL that can produce executable binaries (e.g. AMDGPU), or portable binaries (e.g. SPIR). It is part of [the LLVM compiler infrastructure project](https://llvm.org/), and there is information regarding [OpenCL kernel language support and standard headers](https://clang.llvm.org/docs/UsersManual.html#opencl-features).
+* [libclc](https://libclc.llvm.org/) is a generic and portable implementation of OpenCL builtin function libraries for OpenCL 1.1 - and some functions from later versions of OpenCL can be found there too.
+* [SPIRV-LLVM Translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator) provides a library and the __llvm-spirv__ tool for bidirectional translation between LLVM IR and SPIR-V. 
+* [clspv](https://github.com/google/clspv) compiler and [clvk](https://github.com/kpet/clvk) runtime layer enable OpenCL applications to be executed with Vulkan drivers.
+* [SPIR-V Tools](https://github.com/KhronosGroup/SPIRV-Tools) provide a set of utilities to process SPIR-V binaries including __spirv-opt__ optimizer, __spirv-link__ linker, __spirv-dis__/__spirv-as__ (dis-)assembler, and __spirv-val__ validator.
+
+### Compiling Kernels to SPIR-V
+
+The open source tools can be used to perform full compilation from OpenCL kernel sources into SPIR-V.
+
+<p align="center">
+<br>
+<img src="../images/opencl_to_spirv_tooling.jpg">
+<br> <br>
+  <b>Offline compilation flow for OpenCL kernels into SPIR-V</b>
+<br> <br>
+</p>
+
+#### Examples
+
+If you want to try the above compilation flow for yourself, after installing the tools, you can use the following commands.
+
+##### Compile for OpenCL runtime
+
+__(i)__ Compiling OpenCL C/C++ for OpenCL file into intermediate LLVM IR (for 32 bit targets) formats:
+
+```
+clang -c -target spir -O0 -emit-llvm -o test.bc test.cl
+```
+Using `-cl-std` changes the language version compiled for.
+
+To compile C++ for OpenCL source pass `-cl-std=CLC++` or use the file extension `.clcpp`.
+
+```
+clang -cl-std=CLC++ -c -target spir -O0 -emit-llvm -o test.bc test.cl
+clang -c -target spir -O0 -emit-llvm -o test.bc test.clcpp
+```
+If debugging support is needed then the `-g` flag can be used in the clang invocation:
+
+```
+clang -c test.cl -target spir -o test.bc -g
+```
+
+__Note:__ In clang releases up to 12.0, calling most builtin functions requires the extra flags `-Xclang -finclude-default-header` to be passed to clang. Refer to the release documentation for more details.
+
+__(ii)__ Converting LLVM IR into SPIR-V. 
+
+```
+llvm-spirv test.bc -o test.spv
+```
+
+__Note:__ Converting optimized IR is currently unsupported. However, the stand-alone __spirv-opt__ tool can be used to optimize SPIR-V modules once they have been created.
```

Comment:
I think this is a little too strong.  I frequently convert optimized IR, and there are nice advantages in doing so (when it works).

Reply:
Yes, we do support many features but then some we don't. My worry about advertising it officially is that we still haven't agreed within the dev community that we support this functionality (https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/203). Hence if developers do create bugs related to this we don't guarantee to fix those or we don't guarantee how quickly they will be fixed. So I just want to avoid disappointments. But perhaps we can change the wording to say that this is an **experimental** functionality i.e. expected to work and if not bugs might be fixed but not guaranteed to be fixed.

Reply:
We have discussed it on the Tooling TSG call on Apr 1 and I have captured our decision in the last update.

I would like @svenvh and @AlexeySachkov to proofread this before finalizing the PR.

Reply:
The new wording of this paragraph looks good to me